fix: toggleAttribute polyfill should retain force argument

jmcgavin opened the pull request:
Addresses an issue where `toggleAttribute` would not retain `force`, the second positional argument passed to it.

Fixes https://github.com/webcomponents/polyfills/issues/533

u-b-i-p-i-x-e-l-s commented:
Hey @aomarks and @bicknellr 👋 

Could you please review? Thanks!!

bicknellr commented:
Hi, thanks for sending this! Could you add a test or two? It doesn't seem like we have any for the attribute functions on `Element` yet, but `Element.test.html.js` seems like it would be the right place to put it.

bicknellr commented on packages/scoped-custom-element-registry/test/Element.test.html.js:
```diff
@@ -104,4 +109,41 @@ describe('Element', () => {
+    it('should call setAttribute', () => {
+      const {tagName, CustomElementClass} = getObservedAttributesTestElement(['foo']);
+      customElements.define(tagName, CustomElementClass);
+      const $el = getHTML('<div></div>');
+      $el.innerHTML = `<${tagName}></${tagName}>`;
```
I think this could be simplified to:

```suggestion
      const $el = document.createElement(tagName);
```

here and in the last test.

bicknellr commented on packages/scoped-custom-element-registry/test/Element.test.html.js:
```diff
@@ -104,4 +109,41 @@ describe('Element', () => {
+      const $el = getHTML('<div></div>');
+      $el.innerHTML = `<${tagName}></${tagName}>`;
+
+      $el.firstElementChild.setAttribute('foo', 'bar');
+
+      expect($el.firstElementChild.getAttribute('foo')).to.equal('bar');
```
And then the `.firstElementChild` lookups could be removed down below here and in the other tests:

```suggestion
      $el.setAttribute('foo', 'bar');

      expect($el.getAttribute('foo')).to.equal('bar');
```

bicknellr commented on packages/scoped-custom-element-registry/test/Element.test.html.js:
```diff
@@ -104,4 +109,41 @@ describe('Element', () => {
+
+      expect($el.firstElementChild.hasAttribute('foo')).to.be.false;
+
+      $el.firstElementChild.toggleAttribute('foo', true);
```
I think it would be good to explicitly set the attribute to the empty string before this `.toggleAttribute('foo', true)` to confirm that this call leaves the attribute there if it was already set.

bicknellr commented on packages/scoped-custom-element-registry/test/utils.js:
```diff
@@ -30,6 +30,20 @@ export const getTestElement = () => ({
 });
 
+/**
+ * 
+ * @param {array<string>} observedAttributeNames the names of the attributes you want to observe
+ * @returns {{Klass: typeof HTMLElement, tagName: string}}
```
```suggestion
 * @param {Array<string>} observedAttributeNames the names of the attributes you want to observe
 * @returns {{CustomElementClass: typeof HTMLElement, tagName: string}}
```

bicknellr commented on packages/scoped-custom-element-registry/test/Element.test.html.js:
```diff
@@ -104,4 +109,41 @@ describe('Element', () => {
+
+      expect($el.firstElementChild.hasAttribute('foo')).to.be.false;
+    });
+    it('should call toggleAttribute', () => {
```
```suggestion

    it('should call toggleAttribute', () => {
```

bicknellr commented on packages/scoped-custom-element-registry/test/Element.test.html.js:
```diff
@@ -104,4 +109,41 @@ describe('Element', () => {
+
+      expect($el.firstElementChild.getAttribute('foo')).to.equal('bar');
+    });
+    it('should call removeAttribute', () => {
```
```suggestion

    it('should call removeAttribute', () => {
```

jpzwarte commented:
@jmcgavin any chance you could incorporate the review comments? Do you mind if somebody takes over?

jmcgavin commented:
@jpzwarte Apologies for the delay! I've addressed the comments that were left

jmcgavin commented:
@bicknellr Thank you for the review and my apologies for the tardiness of my response. I've addressed your comments. Let me know if it requires additional changes or tests

bicknellr commented:
Thanks for the updates!

I looked into the failing test and it's seems like an unrelated bug that was exposed by a new `getHTML` call. This line is unconditionally calling `attributeChangedCallback` even though it might be `undefined`:

https://github.com/webcomponents/polyfills/blob/0ab5679e40e1363cceb8e7f4d8279d3bdfebce1f/packages/scoped-custom-element-registry/src/scoped-custom-element-registry.js#L393

The whole outer iteration over `observedAttributes` could be nested in an `if (definition.attributeChangedCallback) { ... }` to work around this:

https://github.com/webcomponents/polyfills/blob/0ab5679e40e1363cceb8e7f4d8279d3bdfebce1f/packages/scoped-custom-element-registry/src/scoped-custom-element-registry.js#L391-L400

Also, could you add a change log entry? ([example here](https://github.com/webcomponents/polyfills/commit/7dc9123fdd371b5f5dce9b1b3d3dba0d4a604521#diff-f8814c39c58e56683d9b22c090880894d2ce76a29ae7e99862d864e885dbb148))

bicknellr reviewed (commented):
One last thing, thanks!

bicknellr reviewed (approved):
Thanks!
